dpnp/backend/extensions/indexing/choose.cpp

```diff
@@ -0,0 +1,453 @@
+            "Unable to allocate packed_shapes_strides device memory");
+    }
+
+    auto packed_chc_offsets = usm_unique_ptr<py::ssize_t>(n_chcs, exec_q);
+    if (packed_chc_offsets.get() == nullptr) {
+        throw std::runtime_error(
+            "Unable to allocate packed_chc_offsets device memory");
+    }
+
+    std::vector<sycl::event> host_task_events;
+    host_task_events.reserve(2);
+
+    std::vector<sycl::event> pack_deps;
+    if (nd == 0) {
+        // special case where all inputs are scalars
+        // need to pass src, dst shape=1 and strides=0
+        // chc_strides already initialized to 0 so ignore
+        std::array<py::ssize_t, 1> scalar_sh{1};
+        std::vector<py::ssize_t> src_strides{0};
+        std::vector<py::ssize_t> dst_strides{0};
+
+        pack_deps = _populate_choose_kernel_params(
+            exec_q, host_task_events, packed_chc_ptrs.get(),
+            packed_shapes_strides.get(), packed_chc_offsets.get(),
+            scalar_sh.data(), sh_nelems, src_strides, dst_strides, chc_strides,
+            chc_ptrs, chc_offsets, n_chcs);
+    }
+    else {
+        auto src_strides = src.get_strides_vector();
+        auto dst_strides = dst.get_strides_vector();
+
+        pack_deps = _populate_choose_kernel_params(
+            exec_q, host_task_events, packed_chc_ptrs.get(),
+            packed_shapes_strides.get(), packed_chc_offsets.get(), src_shape,
+            sh_nelems, src_strides, dst_strides, chc_strides, chc_ptrs,
+            chc_offsets, n_chcs);
+    }
+
+    std::vector<sycl::event> all_deps;
+    all_deps.reserve(depends.size() + pack_deps.size());
+    all_deps.insert(std::end(all_deps), std::begin(pack_deps),
+                    std::end(pack_deps));
+    all_deps.insert(std::end(all_deps), std::begin(depends), std::end(depends));
+
+    sycl::event choose_generic_ev =
+        fn(exec_q, nelems, n_chcs, sh_nelems, packed_shapes_strides.get(),
+           src_data, dst_data, packed_chc_ptrs.get(), src_offset, dst_offset,
+           packed_chc_offsets.get(), all_deps);
+
+    // release usm_unique_ptrs
+    auto chc_ptrs_ = packed_chc_ptrs.release();
+    auto shapes_strides_ = packed_shapes_strides.release();
+    auto chc_offsets_ = packed_chc_offsets.release();
+
+    // free packed temporaries
+    sycl::event temporaries_cleanup_ev = exec_q.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(choose_generic_ev);
+        const auto &ctx = exec_q.get_context();
+
+        using dpctl::tensor::alloc_utils::sycl_free_noexcept;
+        cgh.host_task([chc_ptrs_, shapes_strides_, chc_offsets_, ctx]() {
+            sycl_free_noexcept(chc_ptrs_, ctx);
+            sycl_free_noexcept(shapes_strides_, ctx);
+            sycl_free_noexcept(chc_offsets_, ctx);
+        });
+    });
+
+    host_task_events.push_back(temporaries_cleanup_ev);
+
+    using dpctl::utils::keep_args_alive;
+    sycl::event arg_cleanup_ev =
+        keep_args_alive(exec_q, {src, py_chcs, dst}, host_task_events);
```

Comment:
It doesn't feels right that keeping args alive depends on lifetime of temporaries.
I believe you are doing it in order to combine several independent events into one.

But this could potentially delay deletion of args resulting in increasing of memory consumption.

Though I don't know if it actually happens